### ISSUE TITLE
jsonable_encoder alters json_encoders of Pydantic objects

### ISSUE THREAD

**Issue by slackner** [tiangolo/fastapi#4962]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel, SecretStr
from fastapi.encoders import jsonable_encoder


class Credentials(BaseModel):
    password: SecretStr


ENCODERS = {SecretStr: lambda v: v.get_secret_value() if v is not None else None}

creds = Credentials(password="helloworld")

a = jsonable_encoder(creds)
print(a)  # {'password': '**********'}, as expected
b = jsonable_encoder(creds, custom_encoder=ENCODERS)
print(b)  # {'password': 'helloworld'}, as expected
c = jsonable_encoder(creds)
print(c)  # gives {'password': 'helloworld'}, but should be {'password': '**********'}?


creds = Credentials(password="123456789")

d = jsonable_encoder(creds)
print(d)  # gives {'password': '123456789'}, but should be {'password': '**********'}?
```

### Description

* I'm using `jsonable_encoder` at multiple locations in the code, sometimes with `custom_encoder` and sometimes without.
* My understanding is that passing `custom_encoder` should only affect the current call and not modify any global state.
* What actually happens is that it modifies `__config__.json_encoders`, and as a result, public API endpoints might suddenly start returning secret values that should be filtered.

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

Python 3.9.2

### Additional Context

Using Pydantic version 1.9.1.

**Comment by XingDongZhe** [comment]
I see that it use `__config__.json_encoders` and also pass to next call with `encoder`, so could we use `encoder = encoder.copy()` to avoid  modifies `__config__.json_encoders`?

**Comment by slackner** [comment]
Yes, copying `encoder` before modifying it fixes the issue. Since FastAPI requires at least Python 3.6 merging the dictionaries could also be done with `{**a, **b}`, so an alternative fix would look like this:

```diff
diff --git a/fastapi/encoders.py b/fastapi/encoders.py
index 4b7ffe31..39cc2925 100644
--- a/fastapi/encoders.py
+++ b/fastapi/encoders.py
@@ -52,7 +52,7 @@ def jsonable_encoder(
     if isinstance(obj, BaseModel):
         encoder = getattr(obj.__config__, "json_encoders", {})
         if custom_encoder:
-            encoder.update(custom_encoder)
+            encoder = {**encoder, **custom_encoder}
         obj_dict = obj.dict(
             include=include,  # type: ignore # in Pydantic
             exclude=exclude,  # type: ignore # in Pydantic
```

Note that I'm not sure yet if the change has any negative side effects. Would be good to know to know if there was a reason why `.update()` was used in the first place. :thinking:

**Comment by slackner** [comment]
For reference, the code in question was initially added in https://github.com/tiangolo/fastapi/pull/715. There is no explanation why the pydantic config itself is updated, so it probably was meant to be a copy in the first place.
